### ISSUE TITLE
Stage composition (stages within stages)

### ISSUE THREAD

**Issue by albertofaci** [TNG/JGiven#202]
A bit of background: At the moment we have ~500 acceptance tests in JGiven. These are microservice-level component tests that talk over HTTP with REST endpoints. They use a application running with a DB and some mocks abstract away the other microservices. The test setup is thus quite complex involving setting data up that is not specifically relevant for the test but just _needs to be there_. So far we've used inheritance and a lot of stages in the test with `@ScenarioStage`.

We now feel that we need to compose stages within stages. To have stage methods that are a gateway to a few other methods that live in various stages. And we don't need the sub-stages to appear in the report. We have used inheritance in some instances, but it only really allows to decorate the access to one stage. (Note that these are not nested stages, as they live in a different class).

Product owners are a bit picky about cluttering the report with irrelevant steps and the `@Hidden` functionality feels to much of a workaround for this, and we don't want the stage to be hidden always either.

We basically would like to achieve something like this:
```

public class DrivingTest extends SimpleScenarioTest<DrivingStage> {
  
  @ScenarioStage
  CarStage carStage;

  @Test
  public a_car_can_drive() {
       carStage.given().a_car_exists();
       when().the_user_drives_to("London")
       then().he_gets_to("London")
  }
 
}

public class DrivingStage extends Stage<DrivingStage> {
    
    public DrivingStage the_user_drives_to(@Quoted String location) {
    }

    public DrivingStage he_get_to(@Quoted String location) {
    }
}


public class WheelStage extends Stage<WheelStage>  {
   
   @ProvidedScenarioState
    List<Wheel> wheels; //I want to use this somewhere else
   
   public WheelStage $_wheels_exists(int num) {
       wheels = [something]
   }
}

public class ChassisStage extends Stage<ChassisStage>  {
   public ChassisStage a_chassis_exists() {
   ...
   }
}

//***** here comes the magic
public class CarStage extends Stage<CarStage> {
  
  @ScenarioStage //or @SubStage or some other annotation
  WheelStage wheelStage;

  @ScenarioStage
   ChassisStage chassisStage;

  public CarStage a_car_exists() {
      wheelStage.$_wheels_exists(4);
      chassisStage.a_chassis_exists();
      return self();
  }

}

//Report
Given a car exists
When the user drives to "London"
Then he gets to "London"

```

If I try this, the inner stages (e.g. `WheelStage`) are "outside" the stage framework and `wheels` would not appear populated anywhere else.

My question is this is something that someone has thought of, if there is any technical or logical limitation to do it, and any thoughts. :)

**Comment by janschaefer** [comment]
This is related to #83

**Comment by janschaefer** [comment]
Do you think that this is actually a duplicate of #203?

**Comment by albertofaci** [comment]
@janschaefer  I think so, right?. He wants a `@ScenarioState` field to be set in a different stage, and then be updated in that stage for the next method call

**Comment by albertofaci** [comment]
@janschaefer  I've submitted a pull request that definitely solves **this** ticket :D 
I think the other one too

**Comment by janschaefer** [comment]
I implemented this feature now, based on your pull request #205. It is still experimental and I have to write many additional tests for that, but so far it does not break existing behavior. Some things I have changed regarding to the pull request:
* Have no additional annotation `@ComposedScenarioStage` just use `@ScenarioStage`
* The state of child stages is only update when you call a method on them
* I tried to minimize the necessary reads/updates, there are still more than I wanted to have, but I guess they cannot be reduced much further without losing useful use cases
* I added handling for beforeStage and afterStage method of child stages. This was somehow tricky, especially to have a useful execution of the afterStage methods
